### DISCUSSION TITLE
I can't get context to work without a run time error.

### DISCUSSION THREAD

**Discussion by veggie2u** [solidjs/solid#1277] [Q&A] [answered]
I started the project with the Typescript starter. I was creating my own context and keep getting the `xxxx must have a '[Symbol.iterator]()' method that returns an iterator.` error that I saw a couple of others have. I was getting this error at runtime both localhost and deployed. I found [this playground](https://playground.solidjs.com/?version=1.4.1#NobwRAdghgtgpmAXGGUCWEwBowBcCeADgsrgM4Ae2YZA9gK4BOAxiWGjIbY7gAQi9GcCABM4jXgF9eAM0a0YvADo1aAGzQiAtACsyAegDucAEYqA3EogcuPXgDk4ZXHBGz5ilQDp9EJy5ELKxtuPgEAYQYIF0YABXkAN01xKXcFZTAfZiiYoIgrGXoIZlw0WgheAEFCQgAKAEp+K15BOFwmCoAeAD5mlt5OgAsARm6AdTg1bPheXFpeSKKYqprO-RHeiv7Ox2dXXn1NlrXNyUt8iCFRcVqG3gBebt5avs7F6PF42iSxCWyl+4gYaSI7baqEA6gtbvGJfH7iTb1LC8ES0Zj0eDRLwAczaAFE1HBMbgAEL4ACSIheYCgNRU9Xq5jAkiw4Gg8CQkH8rmoBGInPIVBwdCYrE5ITsAnoZDgMJS0jk6W8+n+H0YeSscAotj4YhkUHoaj4hWKpXKDm5VMaID6+n0Dlohl4ABUiE5mIw0IQ+ABrCCOsi8QyDKB8Bh-couCh8bIfaK8NS0Wg+hNoH1wXhQUQo2hOCAAcj4uEG8idWfwQagjGsEGxfVjzl4wFVuGRAgwHqJwlbKLgneJUgAug9eNLZTkboy+kJ2tXnn1jqD+gMRGgEt0QC2GpI1qv1wvtiZ6Lg5hVyuENMwfYCO0JiSCANRrI8n8pLw-H0+8c+X68gMT9t2IJaM+n5vgeJx9FOEBnMyrKQLAbAtuIvJugKlDUCKLBsBKYS8J2oZwAAymg2LQGoyIES4kQfNGyJjjRUY9pUzCsGQdASAqHgZHQGjaHoGoQHavATL2zBqFWGZQLMbpBoMaDMIMQZoGoai8CYGbFhmZCkdAs4ZrQMjfkwvDONwcBWHyGZyowRFzEII7ACxbEcZ0EAYhpjDdG2vC3l20SIM8jSPLwCS0Jo5hiXe3aBXcIVhZoQ5WMlxTlI2NmMVqfD3PhQiEZl0ZvBOtn2XA3QNJFwnGLwhBQOxsyDNZxV2eZMnELM8xkMQzBoDIFZzG17qet6cmhg1GaxkxymqQmSYpho6YpVqOqyEUJRlBUNlwskjC1IQ8iEGQ1r1mlfDNjkyIyrgMLDjlVHEbpUBqHtB1kF4La8AAPp9vAAAz1La9rVV1fa9RW4amaVmaBjZLUOQNUAJW4hhVjW2JycImYyDIfalLW42hU99CafMEOTVlJ0QI2ZlCIFsNQzlwAHi0LZYMzTRbMuLR+cSdw2pzXMtFdMK3MwwVPMwvAPrwwzQYLLQsuzLQAdF0R80r-TCzkovi-hvBaDLcvy4rAsK2zAuDuc05tB084C0VSziAVuBeNtvyE2oxOAjTcAguzID7bQh3vfJagiFckgQfoGWRllruJDtoLQZHFzLaEq2mhto4yjZ6tbDOtsMbH0a1DHtG4MnVjMoOQA) but was getting the error her too. When you look at this example in the playground, the useCounter still has typescript complaining. For me, it was fine in VsCode, but not when I ran it. I gave up and tried to do this with just jsx, no types using the counter example [from the docs](https://www.solidjs.com/tutorial/stores_context?solved) Now I get `Uncaught (in promise) TypeError: useCounter is not a function or its return value is not iterable`

**Comment by lxsmnsyc** [comment]
> `xxxx must have a '[Symbol.iterator]()' method that returns an iterator.`

This only happens when there's an attempt to destructure a non-array value. It's more likely that the Context isn't detected.

> When you look at this example in the playground, the useCounter still has typescript complaining.

Destructuring without type-elimation causes this. Take note that `useContext` can return undefined.

It would be more helpful to know what your current code looks like.

**Comment by veggie2u** [comment, marked as answer]
> Destructuring without type-elimation causes this. Take note that useContext can return undefined.

Ah! I wasn't understanding that the error message was coming from the fact that useContext could return undefined. So, how do we fix this? I now see that in the [typescript guide](https://www.solidjs.com/guides/typescript#context) it talks about this. It mentions adding a ! to the useFunction. It also says this is bad. Is there a way we could actually get undefined here? Is there a better solution?

I did create a repo with just the SolidJS ts template, and a counter context example [here](https://github.com/veggie2u/solidjs-typescript-counter) I got the error without the !, and it disappears with it.

**Reply by lxsmnsyc** [reply]
> So, how do we fix this?

Check first if `useContext` returns undefined then throw.

```js
const ctx = useContext(MyContext);
if (ctx) {
  return ctx;
}
throw new Error('Missing MyContext');
```

**Reply by TIM-B-R** [reply]
This worked for me. A combination of what was in the repo along with @lxsmnsyc's comment. Thanks both.
